File: modules/doobie/src/main/scala/Queries.scala
/*
 * Copyright 2021 Beyond Scale Group
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edomata.doobie

import _root_.doobie.*
import _root_.doobie.free.connection.ConnectionIO
import _root_.doobie.implicits.*
import _root_.doobie.postgres.implicits.*
import _root_.doobie.util.fragment.Fragment
import cats.data.NonEmptyChain
import edomata.backend.*
import edomata.backend.eventsourcing.AggregateState
import edomata.core.CommandMessage
import edomata.core.MessageMetadata

import java.time.OffsetDateTime
import java.util.UUID

private[doobie] object Queries {
  private def escapeStr(name: PGNamespace) = s""""$name""""
  private def escape(name: PGNamespace) = Fragment.const(escapeStr(name))

  def setupSchema(namespace: PGNamespace): Update0 =
    sql"""create schema if not exists ${escape(namespace)};""".update

  final class Journal[E](naming: PGNaming, codec: BackendCodec[E]) {
    private val table = Fragment.const(naming.table("journal"))
    private val tableStr = naming.table("journal")
    private val payloadOid = Fragment.const(codec.tpe)
    private given Meta[E] = codec.codec
    private val pk = Fragment.const(naming.constraint("journal_pk"))
    private val un = Fragment.const(naming.constraint("journal_un"))
    private val seqnrIdx = Fragment.const(naming.index("journal_seqnr_idx"))
    private val streamIdx = Fragment.const(naming.index("journal_stream_idx"))

    def setup = sql"""
DO $$$$ begin

CREATE TABLE IF NOT EXISTS $table (
  id uuid NOT NULL,
  "time" timestamptz NOT NULL,
  seqnr bigserial NOT NULL,
  "version" int8 NOT NULL,
  stream text NOT NULL,
  payload $payloadOid NOT NULL,
  CONSTRAINT $pk PRIMARY KEY (id),
  CONSTRAINT $un UNIQUE (stream, version)
);

CREATE INDEX IF NOT EXISTS $seqnrIdx ON $table USING btree (seqnr);

CREATE INDEX IF NOT EXISTS $streamIdx ON $table USING btree (stream, version);

END $$$$;
""".update

    final case class InsertRow(
        id: UUID,
        streamId: String,
        time: OffsetDateTime,
        version: SeqNr,
        event: E
    )

    def append(
        n: List[InsertRow]
    ): ConnectionIO[Int] =
      val sql =
        s"""insert into $tableStr ("id", "stream", "time", "version", "payload") values (?, ?, ?, ?, ?)"""
      Update[InsertRow](sql).updateMany(n)

    private val readFields = sql"id, time, seqnr, version, stream, payload"

    def readAll: Query0[EventMessage[E]] =
      sql"select $readFields from $table order by seqnr asc".query

    def readAllAfter(seqNr: Long): Query0[EventMessage[E]] =
      sql"select $readFields from $table where seqnr > $seqNr order by seqnr asc".query

    def readAllBefore(seqNr: Long): Query0[EventMessage[E]] =
      sql"select $readFields from $table where seqnr < $seqNr order by seqnr asc".query

    def readStream(stream: String): Query0[EventMessage[E]] =
      sql"select $readFields from $table where stream = $stream order by version asc".query

    def readStreamAfter(
        stream: String,
        version: Long
    ): Query0[EventMessage[E]] =
      sql"select $readFields from $table where stream = $stream and version > $version order by version asc".query

    def readStreamBefore(
        stream: String,
        version: Long
    ): Query0[EventMessage[E]] =
      sql"select $readFields from $table where stream = $stream and version < $version order by version asc".query
  }

  final class Outbox[N](naming: PGNaming, codec: BackendCodec[N]) {
    private val table = Fragment.const(naming.table("outbox"))
    private val tableStr = naming.table("outbox")
    private given Meta[N] = codec.codec
    private val pk = Fragment.const(naming.constraint("outbox_pk"))

    val setup: Update0 = sql"""
CREATE TABLE IF NOT EXISTS $table(
  seqnr bigserial NOT NULL,
  stream text NOT NULL,
  correlation text NULL,
  causation text NULL,
  payload ${Fragment.const(codec.tpe)} NOT NULL,
  created timestamptz NOT NULL,
  published timestamptz NULL,
  CONSTRAINT $pk PRIMARY KEY (seqnr)
);
""".update

    def markAsPublished(l: NonEmptyChain[Long], time: OffsetDateTime): Update0 =
      sql"""
update $table
set published = $time
where ${Fragments.in(fr"seqnr", l)}
""".update

    val read: Query0[OutboxItem[N]] =
      sql"""
select seqnr, stream, created, payload, correlation, causation
from $table
where published is NULL
order by seqnr asc
""".query

    type ToInsert = (N, String, OffsetDateTime, MessageMetadata)
    def insertAll(items: List[ToInsert]): ConnectionIO[Int] =
      val sql = s"""
insert into $tableStr (payload, stream, created, correlation, causation) values (?, ?, ?, ?, ?)
"""
      Update[ToInsert](sql).updateMany(items)
  }

  final class Snapshot[S](
      naming: PGNaming,
      codec: BackendCodec[S]
  ) {
    private val table = Fragment.const(naming.table("snapshots"))
    private val tableStr = naming.table("snapshots")
    private given Meta[S] = codec.codec
    private val pk = Fragment.const(naming.constraint("snapshots_pk"))

    val setup: Update0 = sql"""
CREATE TABLE IF NOT EXISTS $table (
  id text NOT NULL,
  "version" int8 NOT NULL,
  state ${Fragment.const(codec.tpe)} NOT NULL,
  CONSTRAINT $pk PRIMARY KEY (id)
);
""".update

    type ToInsert = (String, AggregateState.Valid[S])
    def put(l: List[ToInsert]): ConnectionIO[Int] =
      val sql = s"""
insert into $tableStr (id, state, "version") values (?, ?, ?)
on conflict (id) do update
set version = excluded.version,
    state   = excluded.state
         """
      Update[ToInsert](sql).updateMany(l)

    def get(id: String): Query0[AggregateState.Valid[S]] =
      sql"""select state , version from $table where id = $id""".query
  }

  final class Commands(naming: PGNaming) {
    private val table = Fragment.const(naming.table("commands"))
    private val pk = Fragment.const(naming.constraint("commands_pk"))

    val setup: Update0 = sql"""
CREATE TABLE IF NOT EXISTS $table (
  id text NOT NULL,
  "time" timestamptz NOT NULL,
  address text NOT NULL,
  CONSTRAINT $pk PRIMARY KEY (id)
);
""".update

    def count(id: String): Query0[Long] =
      sql"select count(*) from $table where id = $id".query

    def insert(cmd: CommandMessage[?]): Update0 = sql"""
insert into $table (id, address, "time") values (${cmd.id}, ${cmd.address}, ${cmd.time})
""".update
  }

  final class State[S](
      naming: PGNaming,
      codec: BackendCodec[S]
  ) {
    private val table = Fragment.const(naming.table("states"))
    private given Meta[S] = codec.codec
    private val pk = Fragment.const(naming.constraint("states_pk"))

    val setup: Update0 = sql"""
CREATE TABLE IF NOT EXISTS $table (
  id text NOT NULL,
  "version" int8 NOT NULL,
  state ${Fragment.const(codec.tpe)} NOT NULL,
  CONSTRAINT $pk PRIMARY KEY (id)
);
""".update

    import cqrs.AggregateState

    def put(id: String, state: S, version: Long): Update0 =
      sql"""
insert into $table (id, state, "version") values ($id, $state, 1)
on conflict (id) do update
set version = $table.version + 1,
    state   = excluded.state
where $table.version = $version
         """.update

    def get(id: String): Query0[AggregateState[S]] =
      sql"""select state , version from $table where id = $id""".query
  }
}

File: rust/crates/edomata-sqlx/src/queries.rs
#![allow(missing_docs)]
//! SQL statements, mirroring `Queries.scala` of the Skunk and Doobie
//! drivers: same tables, columns, constraints, index names and ordering.
//! Setup DDL comes from `edomata_postgres::ddl`, so the tables created by the
//! driver are exactly the ones `PGSchema` generates for migration tools.

use edomata_postgres::{PGNaming, ddl};

/// Journal table statements.
pub struct JournalQueries {
    pub setup: Vec<String>,
    pub insert: String,
    pub read_all: String,
    pub read_all_after: String,
    pub read_all_before: String,
    pub read_stream: String,
    pub read_stream_after: String,
    pub read_stream_before: String,
}

const JOURNAL_FIELDS: &str = "id, time, seqnr, version, stream, payload";

impl JournalQueries {
    pub fn new(naming: &PGNaming, payload_type: &str) -> Self {
        let t = naming.table("journal");
        Self {
            setup: ddl::journal_statements(naming, payload_type),
            insert: format!(
                "insert into {t} (\"id\", \"stream\", \"time\", \"version\", \"payload\") values ($1, $2, $3, $4, $5)"
            ),
            read_all: format!("select {JOURNAL_FIELDS} from {t} order by seqnr asc"),
            read_all_after: format!(
                "select {JOURNAL_FIELDS} from {t} where seqnr > $1 order by seqnr asc"
            ),
            read_all_before: format!(
                "select {JOURNAL_FIELDS} from {t} where seqnr < $1 order by seqnr asc"
            ),
            read_stream: format!(
                "select {JOURNAL_FIELDS} from {t} where stream = $1 order by version asc"
            ),
            read_stream_after: format!(
                "select {JOURNAL_FIELDS} from {t} where stream = $1 and version > $2 order by version asc"
            ),
            read_stream_before: format!(
                "select {JOURNAL_FIELDS} from {t} where stream = $1 and version < $2 order by version asc"
            ),
        }
    }
}

/// Outbox table statements.
pub struct OutboxQueries {
    pub setup: Vec<String>,
    pub insert: String,
    pub read: String,
    pub mark_published: String,
}

impl OutboxQueries {
    pub fn new(naming: &PGNaming, payload_type: &str) -> Self {
        let t = naming.table("outbox");
        Self {
            setup: ddl::outbox_statements(naming, payload_type),
            insert: format!(
                "insert into {t} (payload, stream, created, correlation, causation) values ($1, $2, $3, $4, $5)"
            ),
            read: format!(
                "select seqnr, stream, created, payload, correlation, causation from {t} where published is NULL order by seqnr asc"
            ),
            mark_published: format!("update {t} set published = $1 where seqnr = ANY($2)"),
        }
    }
}

/// Snapshots table statements.
pub struct SnapshotQueries {
    pub setup: Vec<String>,
    pub get: String,
    pub put: String,
}

impl SnapshotQueries {
    pub fn new(naming: &PGNaming, payload_type: &str) -> Self {
        let t = naming.table("snapshots");
        Self {
            setup: ddl::snapshots_statements(naming, payload_type),
            get: format!("select state, version from {t} where id = $1"),
            put: format!(
                "insert into {t} (id, state, \"version\") values ($1, $2, $3) on conflict (id) do update set version = excluded.version, state = excluded.state"
            ),
        }
    }
}

/// Commands table statements.
pub struct CommandQueries {
    pub setup: Vec<String>,
    pub count: String,
    pub insert: String,
}

impl CommandQueries {
    pub fn new(naming: &PGNaming) -> Self {
        let t = naming.table("commands");
        Self {
            setup: ddl::commands_statements(naming),
            count: format!("select count(*) from {t} where id = $1"),
            insert: format!("insert into {t} (id, address, \"time\") values ($1, $2, $3)"),
        }
    }
}

/// CQRS states table statements.
pub struct StateQueries {
    pub setup: Vec<String>,
    pub get: String,
    pub put: String,
}

impl StateQueries {
    pub fn new(naming: &PGNaming, payload_type: &str) -> Self {
        let t = naming.table("states");
        Self {
            setup: ddl::states_statements(naming, payload_type),
            get: format!("select state, version from {t} where id = $1"),
            put: format!(
                "insert into {t} (id, state, \"version\") values ($1, $2, 1) on conflict (id) do update set version = {t}.version + 1, state = excluded.state where {t}.version = $3"
            ),
        }
    }
}

/// Migrations table and journal rewrite statements.
pub struct MigrationQueries {
    pub create_table: Vec<String>,
    pub select_applied: String,
    pub read_payloads: String,
    pub update_payload: String,
    pub insert_applied: String,
    pub truncate_snapshots: String,
}

impl MigrationQueries {
    pub fn new(naming: &PGNaming) -> Self {
        let migrations = naming.table("migrations");
        let journal = naming.table("journal");
        let snapshots = naming.table("snapshots");
        Self {
            create_table: ddl::migrations_statements(naming),
            select_applied: format!("SELECT \"version\" FROM {migrations}"),
            read_payloads: format!("SELECT id, payload::text FROM {journal} ORDER BY seqnr ASC"),
            update_payload: format!("UPDATE {journal} SET payload = $1::jsonb WHERE id = $2"),
            insert_applied: format!(
                "INSERT INTO {migrations} (\"version\", description) VALUES ($1, $2)"
            ),
            truncate_snapshots: format!("TRUNCATE {snapshots}"),
        }
    }
}

/// `CREATE SCHEMA` for schema-mode naming.
pub fn setup_schema(naming: &PGNaming) -> Vec<String> {
    ddl::schema_statement(naming)
}
